src/toolchain.rs

```diff
@@ -0,0 +1,110 @@
+use std::env;
+use std::ffi::OsString;
+use std::process::{Command, Stdio};
+
+use miette::{IntoDiagnostic, Result, WrapErr};
+use rustc_version::Channel;
+
+/// Check if the target component is installed in the host toolchain, and add
+/// it with `rustup` as needed.
+///
+/// # Note
+/// This function calls `rustc -vV` to retrieve the host triple and the release
+/// channel name.
+#[allow(unused)]
+pub fn check_target_component(component: &str) -> Result<()> {
+    let rustc_meta = rustc_version::version_meta().into_diagnostic()?;
+    let host_target = &rustc_meta.host;
+    let release_channel = &rustc_meta.channel;
+
+    check_target_component_with_rustc_meta(component, host_target, release_channel)
+}
+
+/// Check if the target component is installed in the host toolchain, and add
+/// it with `rustup` as needed.
+pub fn check_target_component_with_rustc_meta(
+    component: &str,
+    host: &str,
+    channel: &Channel,
+) -> Result<()> {
+    // resolve $RUSTUP_HOME, which points to the `rustup` base directory
+    // https://rust-lang.github.io/rustup/environment-variables.html#environment-variables
+    let rustup_home = home::rustup_home().into_diagnostic()?;
+
+    // convert `Channel` enum to a lower-cased string representation
+    let channel_name = match channel {
+        Channel::Stable => "stable",
+        Channel::Nightly => "nightly",
+        Channel::Dev => "dev",
+        Channel::Beta => "beta",
+    };
+
+    // check if the target component is installed in the host toolchain
+    let target_component_is_added = rustup_home
+        .join("toolchains")
+        .join(format!("{channel_name}-{host}"))
+        .join("lib")
+        .join("rustlib")
+        .join(component)
+        .exists();
+
+    if !target_component_is_added {
+        // install target component using `rustup`
+        let pb = crate::progress::Progress::start(format_args!(
+            "Installing target component `{}`...",
+            component
+        ));
+
+        let result = install_target_component(component);
+        let finish = if result.is_ok() {
+            "Target component installed"
+        } else {
+            "Failed to install target component"
+        };
+
+        pb.finish(finish);
+    }
+
+    Ok(())
+}
+
+/// Install target component in the host toolchain, using `rustup target add`
+fn install_target_component(component: &str) -> Result<()> {
```

Comment:
does this need the channel? I think you'll end up installing the target in the wrong channel if you don't pass it from the caller, won't you?

Reply:
oh yeah, I think you are right. let me confirm on my end.

Reply:
@calavera  good catch. I pushed a commit to pass the `+toolchain` so we install the target to the right toolchain.